src/entry.c

```diff
@@ -0,0 +1,361 @@
+entry *entrySetValue(entry *e, sds value) {
+    if (entryHasValuePtr(e)) {
+        sds *value_ref = entryGetValueRef(e);
+        sdsfree(*value_ref);
+        *value_ref = value;
+        return e;
+    } else {
+        entry *new_entry = entryUpdate(e, value, entryGetExpiry(e));
+        return new_entry;
+    }
+}
+
+/* Returns the address of the entry allocation. */
+void *entryAllocPtr(const entry *entry) {
+    char *buf = sdsAllocPtr(entry);
+    if (entryHasValuePtr(entry)) buf -= sizeof(sds);
+    if (entryHasExpiry(entry)) buf -= sizeof(long long);
+    return buf;
+}
+
+bool entryHasEmbeddedValue(entry *entry) {
+    return (!entryHasValuePtr(entry));
+}
+
+/**************************************** Entry Expiry API *****************************************/
+
+/* Returns the entry expiration timestamp.
+ * In case this entry has no expiration time, will return EXPIRE_NONE. */
+long long entryGetExpiry(const entry *entry) {
+    long long expiry = EXPIRY_NONE;
+    if (entryHasExpiry(entry)) {
+        char *buf = sdsAllocPtr(entry);
+        debugServerAssert((((uintptr_t)buf & 0x7) == 0)); /* Test that the allocation is indeed 8 bytes aligned
+                                                           * This is needed since we access the expiry as with pointer casting
+                                                           * which require the access to be 8 bytes aligned. */
+        if (entryHasValuePtr(entry)) buf -= sizeof(sds);
+        buf -= sizeof(long long);
+        expiry = *(long long *)buf;
+    }
+    return expiry;
+}
+
+/* Modify the expiration time of this entry and return a pointer to the (potentially new) entry. */
+entry *entrySetExpiry(entry *e, long long expiry) {
+    if (entryHasExpiry(e)) {
+        char *buf = sdsAllocPtr(e);
+        debugServerAssert((((uintptr_t)buf & 0x7) == 0)); /* Test that the allocation is indeed 8 bytes aligned
+                                                           * This is needed since we access the expiry as with pointer casting
+                                                           * which require the access to be 8 bytes aligned. */
+        if (entryHasValuePtr(e)) buf -= sizeof(sds);
+        buf -= sizeof(expiry);
+        *(long long *)buf = expiry;
+        return e;
+    }
+    entry *new_entry = entryUpdate(e, NULL, expiry);
+    return new_entry;
+}
+
+/* Return true in case the entry has assigned expiration or false otherwise. */
+bool entryIsExpired(entry *entry) {
+    return timestampIsExpired(entryGetExpiry(entry));
+}
+/**************************************** Entry Expiry API - End *****************************************/
+
+void entryFree(entry *entry) {
+    if (entryHasValuePtr(entry)) {
+        sdsfree(entryGetValue(entry));
+    }
+    zfree(entryAllocPtr(entry));
+}
+
+/* Takes ownership of value. does not take ownership of field */
+entry *entryCreate(const_sds field, sds value, long long expiry) {
```

Comment:
passing a ptr to the value sds sends a stronger message for ownership transfer, hence is safer and helps prevent double free or use-after-free.

```suggestion
entry *entryCreate(const_sds field, sds *value_ptr, long long expiry) {
```

Reply:
I understand your point, but it always create the impression that we will place something in the provided reference. I can change that, but there will be some cases were it would "complicate" the calling code, like cases were we do:
```
entry *new_entry = entryCreate(field, sdsdup(value), expiry);
```